ocaml/testsuite/tests/typing-modular-explicit/general.ml

```diff
@@ -0,0 +1,797 @@
+(* TEST
+  expect;
+*)
+
+module type Typ = sig type t end
+
+module type Add = sig type t val add : t -> t -> t end
+
+let id (module T : Typ) (x : T.t) = x
+
+let id2 : (module T : Typ) -> T.t -> T.t =
+  fun (module A : Typ) (x : A.t) -> x
+
+let id3 : (module T : Typ) -> T.t -> T.t =
+  fun (module T) x -> x
+  
+
+[%%expect{|
+module type Typ = sig type t end
+module type Add = sig type t val add : t -> t -> t end
+val id : (module T : Typ) -> T.t -> T.t = <fun>
+val id2 : (module T : Typ) -> T.t -> T.t = <fun>
+val id3 : (module T : Typ) -> T.t -> T.t = <fun>
+|}]
+
+
+let f x y = (id (module Int) x, id (module Bool) y)
+
+[%%expect{|
+val f : Int.t -> Bool.t -> Int.t * Bool.t = <fun>
+|}]
+
+let merge (module T : Typ) x y = (id (module T) x, id (module T) y)
+
+[%%expect{|
+val merge : (module T : Typ) -> T.t -> T.t -> T.t * T.t = <fun>
+|}]
+
+let test_lambda a = (fun (module T : Typ) (x : T.t) -> x) (module Int) a
+
+[%%expect{|
+val test_lambda : Int.t -> Int.t = <fun>
+|}]
+
+
+let alpha_equiv (f : (module A : Add) -> A.t -> A.t) : (module T : Add) -> T.t -> T.t = f
+
+[%%expect{|
+val alpha_equiv :
+  ((module A : Add) -> A.t -> A.t) -> ((module T : Add) -> T.t -> T.t) =
+  <fun>
+|}]
+
+let apply_weird (module M : Typ) (f : (module M : Typ) -> _) (x : M.t) : M.t =
+  f (module M) x
+
+[%%expect{|
+val apply_weird :
+  (module M/1 : Typ) ->
+  ((module M/2 : Typ) -> M/1.t -> M/1.t) -> M/1.t -> M/1.t = <fun>
+|}]
+
+(* Invalid arguments *)
+
+let f x (module M : Typ) (y : M.t) = (x, y)
+
+[%%expect{|
+val f : 'a -> ((module M : Typ) -> M.t -> 'a * M.t) = <fun>
```

Comment:
Yes I noticed it too. I think it is linked with the modes because they impacted how the printing handles parenthesis (this problem didn't appear on the upstream version)

Reply:
Got a patch for this but the test with recursive types is broken now due to #3133